File: csharp/linked-list/LinkedList.cs
public class Deque<T>
{
    private Element first;
    private Element last;

    public void Push(T value)
    {
        var previousLast = last;

        last = new Element { Value = value, Prev = previousLast };
        first = first ?? last;
        
        if (previousLast != null)
        {
            previousLast.Next = last;
        }
    }

    public T Pop()
    {
        var value = last.Value;
        last = last.Prev;

        return value;
    }

    public T Shift()
    {
        var value = first.Value;
        first = first.Next;

        return value;
    }

    public void Unshift(T value)
    {
        var previousFirst = first;

        first = new Element { Value = value, Next = previousFirst };
        last = last ?? first;

        if (previousFirst != null)
        {
            previousFirst.Prev = first;
        }
    }

    private class Element
    {
        public T Value { get; set; }
        public Element Next { get; set; }
        public Element Prev { get; set; }
    }
}

File: fsharp/linked-list/LinkedList.fs
module LinkedList

type Element<'a> =
    { value: 'a
      mutable prev: Element<'a> Option
      mutable next: Element<'a> Option }

type Deque<'a>() =
    let mutable first: Option<Element<'a>> = None
    let mutable last: Option<Element<'a>>  = None

    let addToEmpty newValue = 
        let newElement = { value = newValue; prev = None; next = None }
        last  <- Some newElement
        first <- Some newElement

    member this.pop() =
        match last with
        | None -> failwith "Cannot pop from empty list"
        | Some(x) ->
            last <- x.prev
            x.value

    member this.shift() =
        match first with
        | None -> failwith "Cannot shift from empty list"
        | Some(x) ->
            first <- x.next
            x.value
        
    member this.push(newValue) = 
        match last with
        | None -> addToEmpty newValue
        | Some(x) ->     
            last   <- Some { value = newValue; prev = last; next = x.next }
            x.next <- last            

    member this.unshift(newValue) =        
        match first with
        | None -> addToEmpty newValue
        | Some(x) ->     
            first  <- Some { value = newValue; prev = x.prev; next = first }
            x.prev <- first
            
let mkLinkedList() = Deque()
let push newValue (linkedList: Deque<'T>) = linkedList.push(newValue)
let pop (linkedList: Deque<'T>) = linkedList.pop()
let shift (linkedList: Deque<'T>) = linkedList.shift()
let unshift newValue (linkedList: Deque<'T>) = linkedList.unshift(newValue)
